Fixes #6043 - Reimplement UnixSocket support based on Java 16.

sbordet opened the pull request:
* Introduced new module "jetty-server-unixdomain".
It uses reflection to access the Java 16 Unix-Domain classes to keep compatibility with the other modules and the build.
* Added Jetty module with only HTTP/1.1 support for now (requires review of the modules to reuse them with various connectors).
* Updated documentation to mention UnixDomainServerConnector.
* Updated client libraries to support Unix-Domain.
* Updated PROXY protocol implementation to support Unix-Domain.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #6043 - Reimplement UnixSocket support based on Java 16.

* Introduced new module "jetty-server-unixdomain".
It uses reflection to access the Java 16 Unix-Domain classes to keep compatibility with the other modules and the build.
* Added Jetty module with only HTTP/1.1 support for now (requires review of the modules to reuse them with various connectors).
* Updated documentation to mention UnixDomainServerConnector.
* Updated client libraries to support Unix-Domain.
* Updated PROXY protocol implementation to support Unix-Domain.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-unixdomain-server/src/main/config/etc/jetty-unixdomain.xml:
```diff
@@ -0,0 +1,34 @@
+        <Arg>
+          <Array type="org.eclipse.jetty.server.ConnectionFactory">
+            <Item>
+              <New class="org.eclipse.jetty.server.HttpConnectionFactory">
```
Is `HTTP` the only protocol that we will deliver over UnixDomain sockets?
Can we do `SSL`, `Proxy`, `HTTP/2` ? 

At the very least, this file should be named `jetty-unixdomain-http.xml`

sbordet replied:
@gregw I agree, but the current status for the other implementation is that we have 7 modules for all possible permutations.

I wanted to explore the possibility to separate the "carrier", i.e. IP vs Unix-Domain, from the protocols you attach to it.
Something along the lines of `--add-modules=http2` would implicitly use IP (i.e. `ServerConnector`), while `--add-modules=unix-domain,http2` would use Unix-Domain (i.e. `UnixDomainServerConnector`).

I thought that this could be worked out in a different PR.

gregw replied:
Working out the full topology can be punted to a different PR, but you need to at least start with the right name.    This should be `unixdomain-http` with initially can be complete in itself, but in future might depend on a `unix-domain` module.      If you start with the non protocol module name, then we have less flexibility in future

sbordet replied:
Done.

gregw commented on jetty-unixdomain-server/src/main/config/modules/unixdomain.mod:
```diff
@@ -0,0 +1,36 @@
+[description]
+Enables support for Java 16 Unix-Domain server sockets.
```
```suggestion
Enables support for HTTP over Java 16 Unix-Domain server sockets.
```
rename module to `unixdomain-http` or `http-unixdomain`

sbordet replied:
Done.

gregw reviewed (changes requested):
I'm concerned with hard coding (in config xml) that unix domain is only for raw HTTP.   I would expect that other protocols may be carried (eg PROXY from haproxy), so we should set up to be a bit more flexible.

olamy commented on pom.xml:
```diff
@@ -150,6 +150,7 @@
     <module>jetty-bom</module>
     <module>documentation</module>
     <module>jetty-keystore</module>
+    <module>jetty-unixdomain-server</module>
```
this module can be included only within a jdk16 maven profile.

sbordet replied:
It does not require Java 16 to _build_, only to be used, so there is no need for profiles, not in the build nor in the release.

olamy replied:
in this case you need to remove this https://github.com/eclipse/jetty.project/pull/6522/files#diff-723a454092742efde002c05bbdc642cd7c1a3a4b1d069bffe9ab63fdb1c10bfbR24
```
        <configuration>
          <source>16</source>
          <target>16</target>
          <release>16</release>
        </configuration>
```
because this doesn't build with jdk11.
```
release version 16 not supported
```

sbordet replied:
Ah, good catch!

olamy reviewed (changes requested):
the profile <id>eclipse-release</id> must be updated as well to enforce usage of jdk16 when releasing

sbordet committed:
Removed wrong configuration for Java 16.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Updated tests to be compilable with Java 11 and run on Java 16+.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java:
```diff
@@ -165,12 +167,13 @@ public void configureConnector() throws Exception
+        // Configure IP parameters.
 
-        // The TCP port to listen to.
+        // The IP port to listen to.
```
The port still relates to `TCP`.

lorban commented on documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java:
```diff
@@ -179,6 +182,33 @@ public void configureConnector() throws Exception
+        int selectors = 1;
+
+        // Create a ServerConnector instance.
+        UnixDomainServerConnector connector = new UnixDomainServerConnector(server, 1, 1, new HttpConnectionFactory());
```
Aren't you supposed to use the `acceptors` and `selectors` variables here?

sbordet replied:
Good catch!

lorban commented on jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java:
```diff
@@ -549,24 +549,25 @@ protected boolean removeDestination(HttpDestination destination)
+    protected void newConnection(HttpDestination destination, Promise<Connection> promise)
     {
+        // Multiple threads may access the map, especially with DEBUG logging enabled.
+        Map<String, Object> context = new ConcurrentHashMap<>();
```
Is this a read-only map or a modifiable one? In the former case, an unmodifiable `HashMap` would be preferable.

sbordet replied:
It is a modifiable one, as different `ClientConnectionFactory` may enrich it with additional entries.

lorban commented on jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpClientTransportOverHTTP.java:
```diff
@@ -26,22 +26,24 @@
     public HttpClientTransportOverHTTP()
     {
-        this(Math.max(1, ProcessorUtils.availableProcessors() / 2));
+        this(1);
```
Are you sure you want to change this behavior in this PR?

sbordet replied:
Good point, reverted.

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:
```diff
@@ -221,20 +238,16 @@ public void connect(SocketAddress address, Map<String, Object> context)
 
             boolean connected = true;
-            boolean blocking = isConnectBlocking();
+            boolean blocking = isConnectBlocking() && address instanceof InetSocketAddress;
```
Why is `isConnectBlocking` ignored for non `InetSocketAddress` instances?

sbordet replied:
Because in the code below I cannot call `channel.socket()` if it is a Unix-Domain socket address (it throws).
I would do `channel instanceof SocketChannel`, but that is true for both network and Unix-Domain, despite the fact that calling `socket()` on a `SocketChannel` may throw if it is Unix-Domain. 🤷🏼‍♂️

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:
```diff
@@ -288,9 +301,34 @@ public void accept(SocketChannel channel, Map<String, Object> context)
+        }
+        catch (Throwable x)
+        {
+            if (LOG.isDebugEnabled())
```
I'm not certain silently ignoring the bind address when bind fails is a sensible thing to do. Why do you want to change the existing behavior of throwing when bind fails?

sbordet replied:
Again because for Unix-Domain, binding makes no sense, so if `ClientConnector` is misconfigured, we ignore the binding attempt.

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:
```diff
@@ -288,9 +301,34 @@ public void accept(SocketChannel channel, Map<String, Object> context)
-        channel.socket().setTcpNoDelay(true);
+        try
+        {
+            channel.setOption(StandardSocketOptions.TCP_NODELAY, true);
```
Why forcing this TCP option by default?

sbordet replied:
You are right that this should be configurable, but it's covered by #6372.

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:
```diff
@@ -351,4 +389,78 @@ protected void connectionFailed(SelectableChannel channel, Throwable failure, Ob
+             * destination socket address given as input.</p>
+             * <p>The return type is a pair/record holding the socket channel and the
+             * socket address, with the socket channel not yet connected.
+             * The implementation of this methos must not call
```
Typo: implementation of this **methos** -> **method**

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:
```diff
@@ -351,4 +389,78 @@ protected void connectionFailed(SelectableChannel channel, Throwable failure, Ob
+        /**
+         * <p>A factory for {@link SocketChannelWithAddress} instances.</p>
+         */
+        public interface Factory
```
This mechanism won't work for QUIC as it works with `DatagramChannel` instead of `SocketChannel`.

The `SocketChannelWithAddress` class could be modified to hold a `SelectableChannel` instead of `SocketChannel` but that would not solve the API problem that `ClientConnector` already exposes `SocketChannel` via `configure(SocketChannel channel)` for instance.

This class is sitting between OSI Layer 4 (TCP) and Layer 5 (Sockets, both TCP and Unix Domain), or maybe even overlapping both so I'm not sure I would expose it as an API. Something doesn't look right about it.

joakime replied:
This should probably be a new issue.

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:
```diff
@@ -351,4 +389,78 @@ protected void connectionFailed(SelectableChannel channel, Throwable failure, Ob
+                    {
+                        ProtocolFamily family = Enum.valueOf(StandardProtocolFamily.class, "UNIX");
+                        Class<?> channelClass = Class.forName("java.nio.channels.SocketChannel");
+                        SocketChannel socketChannel = (SocketChannel)channelClass.getMethod("open", ProtocolFamily.class).invoke(null, family);
```
How about `SocketChannel.class.getMethod("open", ProtocolFamily.class).invoke(null, family);`?

sbordet replied:
Much better indeed.

lorban commented on jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java:
```diff
@@ -957,4 +970,47 @@ public void write(Callback callback, ByteBuffer... buffers) throws WritePendingE
+            }
+        }
+
+        private static String toPath(byte[] bytes)
```
This method looks unused.

sbordet replied:
It's used at lines 614 and 616.

sbordet committed:
Updates after reviews.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java:
```diff
@@ -179,6 +182,33 @@ public void configureConnector() throws Exception
+        // The Unix-Domain path to listen to.
+        connector.setUnixDomainPath(Path.of("/tmp/jetty.sock"));
+
+        // The TCP accept queue size.
```
This is unrelated to TCP.

sbordet committed:
Made the mechanism for creating Unix-Domain sockets and addresses private.
This would make easier to fit QUIC in the future.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@lorban I made the mechanism to create a Unix-Domain `ClientConnector` more private, as discussed.

I also updated `FastCGIProxyServlet` to use Unix-Domain when configured so.

sbordet commented:
@gregw @olamy nudge

gregw reviewed (dismissed):
better in that out at this stage!

olamy commented on jetty-fcgi/fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/proxy/FastCGIProxyServletTest.java:
```diff
@@ -216,4 +230,34 @@ public void handle(String target, org.eclipse.jetty.server.Request baseRequest,
+        int maxUnixDomainPathLength = 108;
+        Path path = Files.createTempFile("unix", ".sock");
+        if (path.normalize().toAbsolutePath().toString().length() > maxUnixDomainPathLength)
+            path = Files.createTempFile(Path.of("/tmp"), "unix", ".sock");
```
can be a problem on windows or osx? this test should not run on windows `@DisabledOnOs(OS.WINDOWS)`?

olamy commented on jetty-unixdomain-server/src/test/java/org/eclipse/jetty/unixdomain/server/UnixDomainTest.java:
```diff
@@ -0,0 +1,267 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
```
maybe need `@DisabledOnOs(OS.WINDOWS)`

olamy commented on tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java:
```diff
@@ -906,4 +908,37 @@ public void testDefaultLoggingProviderNotActiveWhenExplicitProviderIsPresent() t
     }
+
+    @Test
+    @EnabledForJreRange(min = JRE.JAVA_16)
```
`@DisabledOnOs(OS.WINDOWS)`

olamy reviewed (changes requested):
might need to disable some tests for windows OS

sbordet committed:
Issue #6043 - Reimplement UnixSocket support based on Java 16.

Replaced unix.socket.tmp with better named jetty.unixdomain.dir property.
Defaulted jetty.unixdomain.dir property to system property user.home under Windows.
Simplified code that runs Unix-Domain tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6043 - Reimplement UnixSocket support based on Java 16.

Fixed Unix-Domain profile for Windows.
Fixed test to make it work on Windows.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

joakime reviewed (approved):
Tested on Windows 10 laptop, works as expected.

sbordet commented:
@olamy this PR now runs fine on Windows too.
There is a new profile to set the Unix-Domain directory for Windows.
Property `unix.socket.tmp` has been removed in favor of `jetty.unixdomain.dir`.
